### PR TITLE
[SegFormer] Add support for segmentation masks with one label

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR makes it possible to fine-tune SegFormer in case you have a mask containing only a single value, i.e. your mask could look like [[255, 0], [0, 255]]. In this case, config.num_labels = 1 and the ignore index is 255.

If this works fine, then we can add it to any other model supported by `AutoModelForSemanticSegmentation`.

### CODE REVIEW

**Comment by dimkoss11** [comment]
How soon will this PR be done?

**Comment by jamilkarabaev** [comment]
when is this going to be closed. Our team would like to use it

**Comment by jamilkarabaev** [comment]
when this becomes active, how would one use it.

thank you nielsrogge for sending for review.

**Comment by jamilkarabaev** [comment]
hello @NielsRogge 

hope I didnt disturb

I tried to peel off the classifier in pytorch and change the output channels to one, then manually compute the loss instead of getting it from the segformer huggingface object, so basically I just got the output and then did a dice loss myself.

So i tried to write bin seg myself kinda, and I started to get a bunch of negative values.

Any idea why that happened, and how I could fix it? I mean, I replaced the last conv2d layer.

**Review by alaradirik — APPROVED** [review]
Looks good to me! But I agree with the commenter about using Dice loss instead of BCEWithLogitsLoss.

**Review by sgugger — APPROVED** [review]
Looks good to me once the loss comment is resolved.

File: src/transformers/models/segformer/modeling_segformer.py [hunk 1]
```diff
@@ -806,15 +806,20 @@ def forward(
...
+            # upsample logits to the images' original size
+            upsampled_logits = nn.functional.interpolate(
+                logits, size=labels.shape[-2:], mode="bilinear", align_corners=False
+            )
+            if self.config.num_labels > 1:
                 loss_fct = CrossEntropyLoss(ignore_index=self.config.semantic_loss_ignore_index)
                 loss = loss_fct(upsampled_logits, labels)
+            elif self.config.num_labels == 1:
+                valid_mask = ((labels >= 0) & (labels != self.config.semantic_loss_ignore_index)).float()
+                loss_fct = BCEWithLogitsLoss(reduction="none")
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/segformer/modeling_segformer.py:818]: Is the Dice loss generally better?

[the remark this answers is not in the document]
Comment (alaradirik) [reply src/transformers/models/segformer/modeling_segformer.py:818]: Dice loss is preferred for tasks that perform binary classification with imbalanced data as the loss converges faster but using BCEWithLogitsLoss is not inaccurate. Many people use a weighted combination of both losses.

[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/segformer/modeling_segformer.py:818]: Is it fine to keep it like this?

[the remark this answers is not in the document]
Comment (jamilkarabaev) [reply src/transformers/models/segformer/modeling_segformer.py:818]: yes, please publish. Although dice loss is quite good for unbalanced segmentation tasks, where foreground is much less than background.

[the remark this answers is not in the document]
Comment (alaradirik) [reply src/transformers/models/segformer/modeling_segformer.py:818]: Yes, I agree with @jamilkarabaev. Dice loss would be nice to have but not required.

[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/segformer/modeling_segformer.py:818]: Ok we'll start with this and can add a config attribute to use "dice" later on.

**Comment by aegonwolf** [comment]
Is this now automatically using dice loss when we set `num_labels = 1`? Maybe I missed it but it seems the documentation doesn't explain it.

**Comment by NielsRogge** [comment]
HHi @aegonwolf,

When config.num_labels = 1, the binary cross-entropy loss is used, as can be seen here: https://github.com/huggingface/transformers/blob/1689aea73346816b936b84932e12b774974e61a6/src/transformers/models/segformer/modeling_segformer.py#L813-L817.

**Comment by NielsRogge** [comment]
@nikolaJovisic the BCE loss uses sigmoid.

**Comment by hungphongtran-pixta** [comment]
@NielsRogge Thank you for your amazing work. However I am still struggling with this task, can you make an example notebook on how to finetune Segformer for 1 class only please? Thank you a lot!

**Comment by nikolaJovisic** [comment]
@NielsRogge modeling_segformer.py script is changed, but modeling_tf_segformer.py seems to be unchanged, I get an error there when I try to train model for 1 class only.

**Comment by pratikkorat26** [comment]
Any update on this task ? Any example notebook ?

**Comment by nikolaJovisic** [comment]
> Any update on this task ? Any example notebook ?

Hi,

Back when I was working on this, the BinaryCrossentropy variant was merged and the issue was closed, but if I remember correctly it was done only for PyTorch version, so I implemented the Tensorflow version myself on a fork of this repository, you can find it on my profile, never made the effort to open a PR tho. Hope this might help. The fine-tuning itself was straightforward after that if I remember correctly.

Best wishes,
Nikola

**Comment by NielsRogge** [comment]
@nikolaJovisic would be great if you could open a PR for it!
